### PR TITLE
Datagen for leap month patterns

### PR DESCRIPTION
**Pull request by Manishearth**
Progress on #3766 / #3865

Previously: https://github.com/unicode-org/icu4x/pull/4212, but this builds on the design from #4217.

### CODE REVIEW

File: components/datetime/src/provider/neo.rs [hunk 1]
```diff
@@ -108,6 +110,29 @@ pub enum MonthSymbolsV1<'data> {
...
+ feature = "datagen",
+ derive(serde::Serialize, databake::Bake),
+ databake(path = icu_datetime::provider::neo),
+)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+pub struct SimpleSubstitutionPattern<'data> {
+    /// The pattern
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub pattern: Cow<'data, str>,
+    /// The index in which to substitute stuff
```
Comment (sffc) [inline components/datetime/src/provider/neo.rs:134]: Nit: State that it is the byte index (not code point index, etc)

Nit: State that there is a weak invariant that the index is less than the byte length of the string

File: provider/datagen/src/transform/cldr/datetime/neo.rs [hunk 2]
```diff
@@ -467,12 +479,34 @@ fn calendar_months(cal: &Value) -> (usize, bool) {
...
+            Context::Format,
+            "numeric months only found for Context::Format"
+        );
+        let Some(ref patterns) = data.month_patterns else {
+            panic!("No month_patterns found but numeric months were requested for {calendar} with {langid}");
+        };
+        let pattern = patterns.get_symbols(context, length);
+        let leap = &pattern.leap;
+        let Some(index) = leap.find("{0}") else {
+            panic!("Calendar {calendar} for {langid} has leap patterh {leap}, which does not contain {{0}} placeholder");
```
Comment (sffc) [inline provider/datagen/src/transform/cldr/datetime/neo.rs:500]:
```suggestion
            panic!("Calendar {calendar} for {langid} has leap pattern {leap}, which does not contain {{0}} placeholder");
```

**Review by sffc — DISMISSED** [review]

**Review by sffc — COMMENTED** [review]
.

**Comment by Manishearth** [comment]
autosubmit=on

**Review by sffc — APPROVED** [review]

**Pull request merged**